969: Add trackpad mode preference, trackpad gesture support

coreh opened the pull request:
Fixes #969, #2966.

👋 This PR aims to improve the usability of TrenchBroom for trackpad users:

- Introduces a new “Trackpad Mode” preference, for tailoring the input for trackpads (_particularly useful_ on macOS, but also a usability enhancement for Windows and Linux)
- Adds support for Pinch and Rotate gestures in the trackpad, by extending the existing support for scrolling:
  - Adds new `WheelAxis::Pinch` and `WheelAxis::Rotate` variants
  - Adds new `scrollZ()` and `scrollW()` axes on `InputState`
  - Based on [`QNativeGestureEvent`](https://doc.qt.io/qt-5/qnativegestureevent.html)(As of Qt 5, this is only fired on macOS, but Qt 6 also supports it under Wayland.)
- Adds special trackpad logic for the three camera tools (2D, 3D and UV)
  - Zoom level is honored
  - Inverse scroll direction preference is honored
  - Sensitivity preference is honored
- Fixes macOS build instructions by explicitly referencing Qt5 (6 is now the default in Homebrew)

## Control Scheme

### 3D Viewport

Input | Action
-|-
Two Finger Scroll (either direction) | Pan
Alt + Two Finger Scroll (either direction) | Orbit
Two Finger Pinch | Pan (backward/forward)
Shift + Two Finger Pinch | Zoom
Two Finger Rotate | Orbit
Shift + Two Finger Scroll (up/down) | Pan (backward/forward)

### 2D and UV Viewport

Input | Action
-|-
Two Finger Scroll (either direction) | Pan
Two Finger Pinch | Zoom
Shift + Two Finger Scroll (up/down) | Zoom

## Screenshot

<img width="1012" alt="image" src="https://github.com/user-attachments/assets/c70e86db-1678-4f5e-a384-08446d0cf0e2">

coreh commented on common/src/View/CameraTool3D.cpp:
```diff
@@ -154,6 +154,41 @@ void CameraTool3D::mouseScroll(const InputState& inputState)
+    if (orbit)
+    {
+      m_camera.orbit(
+        vm::vec3f{m_camera.defaultPoint(inputState.pickRay())},
```
Originally I was applying the same hit testing used for the drag tracker here, only without storing the orbit point across the whole interaction. It was producing weird results (since another brush could appear "in front" of the orbit point and capture the orbit) so I removed it.

We can store the orbit point on the `CameraTool3D` instance, but I'm worried whether or not that is clean architecturally. Let me know if you'd like me to do that, or something else.

kduske replied:
Yeah, this is a bit unfortunate. I introduced `DragTracker` to avoid storing any data associated with gestures in the tools themselves. I suggested introducing a notion of gestures in the previous commit. I think we might then also want to generalize `DragTracker` to `GestureTracker` and generalize its interface to have `start/update/stop` semantics. Then we could use that for trackpad gestures as well as dragging gestures.

kduske commented on common/src/Preferences.cpp:
```diff
@@ -223,6 +223,7 @@ Preference<float> CameraPanSpeed("Controls/Camera/Pan speed", 0.5f);
 Preference<bool> CameraPanInvertH("Controls/Camera/Invert horizontal pan", false);
 Preference<bool> CameraPanInvertV("Controls/Camera/Invert vertical pan", false);
 Preference<bool> CameraMouseWheelInvert("Controls/Camera/Invert mouse wheel", false);
+Preference<bool> CameraTrackpadMode("Controls/Camera/Trackpad mode", false);
```
I was wondering why it is necessary to introduce a new preference for this. Zoom and rotate events will always come from the trackpad, and scroll events seem to have their source set to `Qt::MouseEventSynthesizedBySystem`, at least on macOS.

So if we know that an event came from the trackpad, would that be enough to avoid introducing this preference?

coreh replied:
Besides enabling support for the gestures, this preference also changes the behavior of the horizontal and vertical scroll events: With it disabled, vertical scrolling is still used for panning backward/forward (and zooming), so the workflow of existing users with mice is unaffected. With it enabled, scrolling horizontally and vertically are changed to behave as horizontal and vertical pans.

<img width="789" alt="image" src="https://github.com/user-attachments/assets/42050c40-e2cf-4614-8fcc-52a50824b81b">

Blender has a similar toggle that does roughly the same behavior change, which is where I copied the idea from.

Ideally we would be able to know if the scroll event came from a mouse or trackpad, and switch automatically, but `QWheelEvent` doesn't seem to expose that information, unfortunately. (And `QNativeGestureEvent` is not fired for two-finger scroll gestures)

kduske replied:
> Ideally we would be able to know if the scroll event came from a mouse or trackpad, and switch automatically, but `QWheelEvent` doesn't seem to expose that information, unfortunately. (And `QNativeGestureEvent` is not fired for two-finger scroll gestures)

But we do know: Zoom and rotate events will always come from the trackpad, and scroll events seem to have their source set to `Qt::MouseEventSynthesizedBySystem` (when they originate at the trackpad), at least on macOS.

coreh replied:
Is it possible to get a `QMouseEvent` from a `QWheelEvent`? The docs only list the `source()` getter for `QMouseEvent`

kduske replied:
I think you missed it, it's here: https://doc.qt.io/qt-5/qwheelevent.html#source

coreh replied:
Oh, I was looking at the _wrong_ `QWheelEvent`: https://doc.qt.io/qt-5/qt3dinput-qwheelevent.html 🥲

Looks like there's a different one inside of `Qt3DInput`

kduske commented on common/src/View/InputEvent.cpp:
```diff
@@ -193,6 +193,12 @@ std::ostream& operator<<(std::ostream& lhs, const MouseEvent::WheelAxis& rhs)
   case MouseEvent::WheelAxis::Vertical:
     lhs << "Vertical";
     break;
+  case MouseEvent::WheelAxis::Pinch:
```
It feels not quite right to introduce these fake wheel axes. Maybe we could rename the enum to something more generic like `GestureType`?

kduske replied:
Or is it common practice to treat these events as if they were associated with virtual "axes"?

coreh replied:
There's _some_ precedent in HTML/DOM: The [`deltaZ`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaZ) axis of `WheelEvent` is used for pinching gestures. (I don't think rotation trackpad gestures are supported on the web, at least with the standardized APIs)

If you feel like it's a hack, I can introduce a separate event type, distinct from scroll specifically for them. (But we'd probably still consolidate them in `InputState`, correct?)

kduske commented on common/src/View/InputState.h:
```diff
@@ -111,6 +113,14 @@ class InputState
    * Number of "lines" to scroll vertically.
    */
   float scrollY() const;
+  /**
```
If we decide to generalize the `WheelAxis` enum, we should also choose clearer names such as `pinchAmount` and `rotateAmount` or something like that.

coreh replied:
Yeah, that's probably cleaner. I'll rename them.

kduske commented on common/src/View/RenderView.cpp:
```diff
@@ -185,6 +185,22 @@ void RenderView::wheelEvent(QWheelEvent* event)
+  // so we must override the main event handler to handle it.
+  if (event->type() == QEvent::NativeGesture)
+  {
+    QNativeGestureEvent* gestureEvent = static_cast<QNativeGestureEvent*>(event);
```
Please use `auto` whenever possible. I know it's controversial, but it's the style that I use now, and I clean up every file that I touch to follow this practice.

kduske commented on common/src/View/CameraTool2D.cpp:
```diff
@@ -70,6 +87,29 @@ static void zoom(
+                         + m_camera.up() * factor * inputState.scrollY();
+      m_camera.moveBy(delta * moveSpeed(m_camera, false));
+    }
+    return;
```
I think I would prefer an `else` over the early return.

coreh replied:
I did the early `return` mostly to avoid touching the other lines with the indentation change, if that's not an issue I can use `else`.

kduske commented on common/src/View/CameraTool2D.cpp:
```diff
@@ -36,7 +36,7 @@ namespace View
   const Renderer::OrthographicCamera& camera, const float speed)
 {
-  return speed * vm::min(1.0f, 1.0f / camera.zoom());
+  return speed / vm::max(0.001f, camera.zoom());
```
I don't understand this change, could you explain this please?

coreh replied:
I had originally copied over and adapted the code from `CameraTool3D`, where it was:

```cpp
return speed * vm::min(1.0f, camera.zoomedFov() / camera.fov());
```

The ratio of the zoom / original FoV produces a value that's > 1 when zoomed out and < 1 when zoomed in. So when zoomed in we want the pan to be slower, and when zoomed out to be faster.

After fiddling with it for a bit I realized that this was backwards for the 2D camera, since the `camera.zoom()` value will be < 1 when zoomed out and > 1 when zoomed in, so I switched the `*` to a `/` to invert it. The `vm::max()` is there simply as a safeguard to prevent producing Infinity / NaN values if the zoom value is ever 0 (which shouldn't happen)

kduske commented on common/src/View/UVCameraTool.cpp:
```diff
@@ -90,7 +90,6 @@ void UVCameraTool::mouseScroll(const InputState& inputState)
     {
       const float zoomFactor =
         1.0f + inputState.scrollZ() * factor + inputState.scrollY() / 50.0f * factor;
-      const auto mousePos = vm::vec2f{inputState.mouseX(), inputState.mouseY()};
```
If this is a fixup for a previous commit, then please create a `fixup!` commit so that it gets squashed.

coreh replied:
I'm not super familiar/confident with (auto)squash workflows, I'll manually rebase and squash it into the right commit

kduske reviewed (commented):
Thank you for this PR! It was a pleasure to review because the history was very clean and easy to follow. Thank you for taking the effort!

I have a bigger design suggestion: I think we need to introduce the notion of "gesture" and generalize the existing infrastructure for mouse drag gestures to also support trackpad gestures. You can find more info in the comments.

If you feel that's out of scope for this PR, then I can take on the redesign work. If you want to do it, I would suggest we discuss and clarify the details before you start working on it.

What do you think?

coreh commented:
Hey @kduske 👋 Thanks for the speedy and throughout review! I'll read through and reply to/address the comments one by one.

As for the larger architectural change into a gesture system, I can try to understand the current drag tracker implementation a little bit better and come up with a generalized design that works with drag and trackpad gestures.

One complication from that is that I believe it wouldn't be fully cross platform without some time or mouse move heuristics, even in a potential scenario where the project is migrated to Qt 6: I believe only on macOS the trackpad emits start/end events for the gestures.

kduske commented:
> As for the larger architectural change into a gesture system, I can try to understand the current drag tracker implementation a little bit better and come up with a generalized design that works with drag and trackpad gestures.

I can also do this in a separate PR, and then you could rebase your PR onto that and use the generalized gesture support to implement your trackpad features. What do you think? I have some time in the next days, so it shouldn't take too long.

coreh commented:
Okay, addressed most of the feedback, and force-pushed a rebased version of the branch with all the fix-ups integrated.

Re: Renaming `MouseEvent::WheelAxis` to `MouseEvent::GestureType` to make it more general, I see three main possible approaches, with various degrees of disruption, so wanted to know what would be preferred:

1. Keep the existing `MouseEvent::Type::Scroll` name, and also all scroll related names (`mouseScroll()`, `processScroll()`, `scroll()`, `scrollDistance`) exactly as-is, and just reuse them for gesture processing, accepting the inconsistent names. _(Least disruptive option)_
2. Rename the existing `MouseEvent::Type::Scroll` (and all scroll-related names) to `MouseEvent::Type::Gesture` (and gesture-related names) for full consistency. `WheelAxis::Horizontal`/`WheelAxis::Vertical` would become `GestureType::HorizontalScroll` and `GestureType::VerticalScroll`. This will imply updating names in a lot of places, but the logic will remain otherwise the same. _(Medium disruptivity option)_
3. Introduce a separate `MouseEvent::Type::Gesture` event, and a separate `MouseEvent::GestureType` enum. Add new handlers/processors/toolbox chain. Add additional parameters to the `MouseEvent` constructor. Add new collate logic. _(Most disruptive option)_

I'm personally kinda scared of Option 3, as that will be very complex and will risk regressing existing functionality. I'd prefer either going with Option 1 or not renaming `WheelAxis` at all.

--- 

Re: Using `source()` for auto-detection, I'll do some experiments with it to figure out how reliable it is. Since it seems to be macOS-only, we'd still need to provide some manual fallback for Windows/Linux users with precision trackpads. There's also the Magic Mouse, which is halfway between a Mouse and a Trackpad, and the user might want to choose what behavior to get.

One thing that I just realized is that I use [Mac Mouse Fix](https://macmousefix.com) and that might interfere with the detection. Other similar tools exist, (e.g. [Better Touch Tool](https://folivora.ai), [Karabiner](https://karabiner-elements.pqrs.org)) I'll try to test them as well to see what Qt reports.

Perhaps instead of a checkbox, what we actually want is a dropdown with three possible values? (Auto-detect, Trackpad, Mouse)?

---

Finally, re: the refactor of the existing drag system to also accommodate for the stateful orbit gesture, if you have the bandwidth to do that, I can update this PR to leverage it once it's ready. It will definitely come out better than if I tried to implement it myself 😄

kduske commented:
I won't be able to review this and do the "gesture" generalization until Monday. I'll look into it then.

Regarding your options, I would actually prefer option 3, because that is would I would have done if I had implemented the trackpad support in the first place.

kduske commented:
>     3. Introduce a separate `MouseEvent::Type::Gesture` event, and a separate `MouseEvent::GestureType` enum. Add new handlers/processors/toolbox chain. Add additional parameters to the `MouseEvent` constructor. Add new collate logic. _(Most disruptive option)_

I'm looking into this, but a question that came up is whether we need to support simultaneous trackpad gestures, i.e. zoom and rotate. I think we do, and I think the right way would be to have one `TrackpadGestureTracker` type and treat pinch, zoom, etc. as one gesture with different inputs, i.e. it will have an `update` function that takes the input state, and looks at the state to determine what has changed (rotate angle, zoom value, ...). I would rather not split these gestures into different tracker classes, because I imagine that a tool that reacts to a zoom gesture will also want to react to rotate.

> Re: Using `source()` for auto-detection, I'll do some experiments with it to figure out how reliable it is. Since it seems to be macOS-only, we'd still need to provide some manual fallback for Windows/Linux users with precision trackpads. There's also the Magic Mouse, which is halfway between a Mouse and a Trackpad, not sure what that would be detected as, and the user might want to choose what behavior to get.

> One thing that I just realized is that I use [Mac Mouse Fix](https://macmousefix.com) and that might interfere with the detection. Other similar tools exist, (e.g. [Better Touch Tool](https://folivora.ai), [Karabiner](https://karabiner-elements.pqrs.org)) I'll try to test them as well to see what Qt reports.

Did you get any results?

> Finally, re: the refactor of the existing drag system to also accommodate for the stateful orbit gesture, if you have the bandwidth to do that, I can update this PR to leverage it once it's ready. It will definitely come out better than if I tried to implement it myself 😄

Yeah, I'm working on this.

coreh commented:
> Regarding your options, I would actually prefer option 3, because that is would I would have done if I had implemented the trackpad support in the first place.

Okay, I'll try to implement pinch/rotate gestures as an extra mouse event type, separate from scrolling. As it will be a large-ish change, I'll probably separate it into its own PR then.

> I'm looking into this, but a question that came up is whether we need to support simultaneous trackpad gestures, i.e. zoom and rotate. I think we do, and I think the right way would be to have one TrackpadGestureTracker type and treat pinch, zoom, etc. as one gesture with different inputs, i.e. it will have an update function that takes the input state, and looks at the state to determine what has changed (rotate angle, zoom value, ...). I would rather not split these gestures into different tracker classes, because I imagine that a tool that reacts to a zoom gesture will also want to react to rotate.

We can support them simultaneously, and I agree it makes sense for the tool/gesture/state code to be the same. However, we won't be able to able fully test the combined rotate+pinch behavior with any of the current backends, as I believe neither macOS or Wayland support a combination of the two gestures at the same. (What happens if you try to use both is that they alternate between each other: There's probably some filtering/heuristic going on on the OS/driver level, as it's probably impossible to do a perfect pinch without rotation or a perfect rotation without pinching in real life.). If we were to support touch screens, then a custom handler for both would be possible, but it would likely take place outside of the existing `MouseEvent` logic (unless we emulated mouse events from the touch gestures)

> Did you get any results?

Not yet, I'll do the experiments as part of the `MouseEvent` gesture implementation. If it works out I will also extend MouseEvent with an extra property for trackpad/mouse detection.

> Yeah, I'm working on this.

👍 In the mean time, I will do the work on the `MouseEvent` gesture on the separate PR, as it will likely be fairly parallel to that. Once both changes are integrated, I will refactor this one to make use of the newly added infrastructure.

kduske commented:
Please hold off on the MouseEvent work for a bit. I'll provide a new GestureEvent that you can then use. It'll be done in a few days.

kduske commented:
Hey @coreh I have implemented basic support for trackpad gestures here: https://github.com/TrenchBroom/TrenchBroom/pull/4596

To implement your features, you would implement `ToolController::acceptGesture` and return a instance of `GestureTracker`. It's pretty similar to how mouse dragging is implemented in TB.

Currently, I haven't added a change where we would treat trackpad scrolling as a gesture. I can do this if we are sure that the event source can be reliably used to determine if the event came from the trackpad or a mouse wheel. If that isn't the case, we can add a preference for this, but I would still implement this as a separate trackpad gesture with a new gesture type `Scroll`, unless there are other problems that I cannot see right now.

kduske commented:
Could you try to implement support for zoom and rotate based on that branch, and then we discuss the scrolling issue?

kduske commented:
One more thought: We can't turn scroll events into a gesture anyway because we don't get the begin/end events. So scrolling is not a gesture. I'll add more data to `InputState` to make it possible to differentiate where the scroll event came from - or alternatively, I might add a new event handler for trackpad scroll events that is separate from mouse scroll events.

Another question that came up is whether we want to use the panning gesture instead of scroll events.

coreh commented:
Oops, sorry! Ended up not seeing your replies to this thread. I'll read through them carefully and port the rotation and zooming to use your new gesture implementation, and open a new PR. Then we can circle back to the panning.

kduske commented:
@coreh Did you get anywhere? Let me know if you ran into problems. I'll gladly help out!

coreh commented:
@kduske Hey, sorry for the radio silence. Finally have some free time/bandwidth this weekend to switch gears to work on this again, will push something in a bit

kduske commented:
Awesome! 👍

coreh commented:
@kduske Opened a new draft PR (#4606), maybe we can close this one, and continue the discussion there.

It was not as smooth sailing as I had expected, mainly because Qt / macOS are doing something weird by interleaving multiple gestures, which I didn't expect.

kduske commented:
Closed in favor of https://github.com/TrenchBroom/TrenchBroom/pull/4606
